shared/src/econ/mod.rs

```diff
@@ -1,7 +1,384 @@
+    #[inline]
+    fn add_assign(&mut self, other: &TokenAmount) {
+        self.atto += &other.atto;
+    }
+}
+
+// Implements Sub for all combinations of value/reference receiver and parameter.
+macro_rules! impl_sub {
+    ($(impl<$($a:lifetime),*> Sub<$Other:ty> for $Self:ty;)*) => {$(
+        impl<$($a),*> Sub<$Other> for $Self {
+            type Output = TokenAmount;
+
+            #[inline]
+            fn sub(self, other: $Other) -> TokenAmount {
+                // automatically match value/ref
+                let TokenAmount { atto: x, .. } = self;
+                let TokenAmount { atto: y, .. } = other;
+                TokenAmount {atto: x - y}
+            }
+        }
+    )*}
+}
+impl_sub! {
+    impl<> Sub<TokenAmount> for TokenAmount;
+    impl<'b> Sub<&'b TokenAmount> for TokenAmount;
+    impl<'a> Sub<TokenAmount> for &'a TokenAmount;
+    impl<'a, 'b> Sub<&'b TokenAmount> for &'a TokenAmount;
+}
+
+impl SubAssign<TokenAmount> for TokenAmount {
+    #[inline]
+    fn sub_assign(&mut self, other: TokenAmount) {
+        self.atto -= &other.atto;
+    }
+}
+
+impl<'a> SubAssign<&'a TokenAmount> for TokenAmount {
+    #[inline]
+    fn sub_assign(&mut self, other: &TokenAmount) {
+        self.atto -= &other.atto;
+    }
+}
+
+// Implements Mul for all combinations of value/reference receiver and 32/64-bit value.
+macro_rules! impl_mul {
+    ($(impl<$($a:lifetime),*> Mul<$Other:ty> for $Self:ty;)*) => {$(
+        impl<$($a),*> Mul<$Other> for $Self {
+            type Output = TokenAmount;
+
+            #[inline]
+            fn mul(self, other: $Other) -> TokenAmount {
+                // automatically match value/ref
+                let TokenAmount { atto: x, .. } = self;
+                // let TokenAmount { atto: y, .. } = other;
+                TokenAmount {atto: x * other}
+            }
+        }
+    )*}
+}
+impl_mul! {
+    impl<> Mul<u32> for TokenAmount;
+    impl<'a> Mul<u32> for &'a TokenAmount;
+    impl<> Mul<i32> for TokenAmount;
+    impl<'a> Mul<i32> for &'a TokenAmount;
+    impl<> Mul<u64> for TokenAmount;
+    impl<'a> Mul<u64> for &'a TokenAmount;
+    impl<> Mul<i64> for TokenAmount;
+    impl<'a> Mul<i64> for &'a TokenAmount;
+}
+
+impl MulAssign<TokenAmount> for TokenAmount {
+    #[inline]
+    fn mul_assign(&mut self, other: TokenAmount) {
+        self.atto *= &other.atto;
+    }
+}
+
+impl<'a> MulAssign<&'a TokenAmount> for TokenAmount {
+    #[inline]
+    fn mul_assign(&mut self, other: &TokenAmount) {
+        self.atto *= &other.atto;
+    }
+}
+
+// Only a single div/rem method is implemented, rather than the full Div and Rem traits.
+// Division isn't a common operation with money-like units, and deserves to be treated carefully.
+impl TokenAmount {
+    #[inline]
+    pub fn div_rem(&self, other: impl Into<BigInt>) -> (TokenAmount, TokenAmount) {
+        let (q, r) = self.atto.div_rem(&other.into());
+        (TokenAmount { atto: q }, TokenAmount { atto: r })
+    }
+}
```

Comment:
I have to wonder if we even _need_ this, or if users should be forced to call `.atto()` and do their own thing.

Reply:
I think division is just likely enough to be relevant to some actors that it was worth adding this one method for it.

Reply:
Fair enough.